Fixes SI-6189.

axel22 opened the pull request:
Disable @static for the REPL code.
The problem was that there are no companion classes generated for objects
that contain the top-level declarations in the REPL.
When there is no companion class, the selecting a @static field will translate
to a getter call, instead of to a field load.

Review by @paulp

Commit:
Fixes SI-6189.

Disable @static for the REPL code.
The problem is that there are no companion classes generated for objects
that contain the top-level declarations in the REPL.
When there is no companion class, the selecting a @static field will translate
to a getter call, instead of to a field load.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/206/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/912/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/206/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/912/

paulp commented on src/compiler/scala/tools/nsc/backend/icode/GenICode.scala:
```diff
@@ -880,12 +880,16 @@ abstract class GenICode extends SubComponent  {
         if !ctx.method.symbol.isStaticConstructor 
-        && fun.symbol.isAccessor && fun.symbol.accessed.hasStaticAnnotation =>
+        && fun.symbol.isAccessor && fun.symbol.accessed.hasStaticAnnotation
+        && qual.tpe.typeSymbol.orElse(fun.symbol.owner).companionClass != NoSymbol =>
```
qual.tpe.typeSymbol.orElse(fun.symbol.owner) looks a bit sketchy, but I don't have anything concrete to propose.

paulp commented:
Does this commit fix SI-6236 as well? Either way, this LGTM for now.  I have some work on @static which uncovered some other bugs but I need to organize it.

axel22 commented:
```
package p {
  object x {
    object x {
      @scala.annotation.static val foo: Int = 353
    }
  }
}
```

No, it does not solve SI-6236 - for some strange reason I'm trying to figure out, in separate compilation, although the `static` field gets generated, it's not loaded into the list of members symbol for `p.x$x`.

Commit:
Fixes SI-6236.

In separate compilation runs, the static field symbol in the
companion class of an object was not being recreated.
Given that the singleton object was compiled separately,
the static field symbol will be recreated on demand.

axel22 commented:
Ok, I've added a fix for SI-6236.

Maybe we should merge this into 2.10.x, given that all tests pass.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/222/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/930/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/930/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/222/

paulp commented:
I have a suspicion we aren't done with @static.  I cherry-picked these commits into 2.10.x.

axel22 commented:
Ok, thanks!
If you observe any other problems with `@static`, please let me know!
